src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Publish.targets

```diff
@@ -485,30 +485,36 @@ Copyright (c) .NET Foundation. All rights reserved.
                       PlatformLibraryName="$(MicrosoftNETPlatformLibrary)"
                       FilesToSkip="@(_ConflictPackageFiles);@(_PublishConflictPackageFiles)"
                       CompilerOptions="@(DependencyFileCompilerOptions)"
-                      PrivateAssetsPackageReferences="@(PrivateAssetsPackageReference)"
+                      ExcludeFromPublishPackageReferences="@(ExcludeFromPublishPackageReference)"
                       TargetManifestFileList="@(TargetManifestFileList)"
                       IsSelfContained="$(SelfContained)" />
 
   </Target>
 
   <!--
     ============================================================
-                                        ComputePrivateAssetsPackageReferences
+                                        ComputeExcludeFromPublishPackageReferences
 
-    Builds up the @(PrivateAssetsPackageReference) item by looking for @(PackageReference) items with 
+    Builds up the @(ExcludeFromPublishPackageReference) item by looking for @(PackageReference) items with 
     <PrivateAssets> metadata.
     ============================================================
     -->
   <PropertyGroup>
-    <ComputePrivateAssetsPackageReferences Condition="'$(ComputePrivateAssetsPackageReferences)' == ''">true</ComputePrivateAssetsPackageReferences>
+    <ComputeExcludeFromPublishPackageReferences Condition="'$(ComputeExcludeFromPublishPackageReferences)' == ''">true</ComputeExcludeFromPublishPackageReferences>
   </PropertyGroup>
 
-  <Target Name="ComputePrivateAssetsPackageReferences"
-          Condition="'$(ComputePrivateAssetsPackageReferences)' == 'true'">
+  <Target Name="ComputeExcludeFromPublishPackageReferences"
+          Condition="'$(ComputeExcludeFromPublishPackageReferences)' == 'true'">
 
     <ItemGroup>
-      <PrivateAssetsPackageReference Include="@(PackageReference)"
-                                     Condition="'%(PackageReference.PrivateAssets)' == 'All'" />
+      
+      <!-- PrivateAssets="All" means exclude from publish, unless ExcludeFromPublish metadata is specified separately -->
+      <PackageReference Update="@(PackageReference)" 
+                        Publish="false"
+                        Condition="('%(PackageReference.PrivateAssets)' == 'All') And ('%(PackageReference.Publish)' == '')"/>
+      
+      <ExcludeFromPublishPackageReference Include="@(PackageReference)"
+                                          Condition="('%(PackageReference.Publish)' == 'false')" />
```

Comment:
Maybe both the itemgroup and the target that computes it should be "private" (underscore-prefixed) because adding to manipulating the item group directly doesn't give you anything more than setting the metadata, does it?